### PR TITLE
[Core] Optionally delete `Ref` `nullptr` comparisons

### PR DESCRIPTION
**Pull request by AThousandShips**
Adds an optional (default true) compile option to allow comparing `Ref` to `nullptr` to ensure correct use

While using `DISABLE_DEPRECATED` would cover the compatibility of this we can't catch any errors in most builds with that, as we only do a build for Linux/BSD with that flag, and then not an editor build

Any suggestions for naming is welcome if this is not optimal 

Followup to:
* https://github.com/godotengine/godot/pull/96292

### CODE REVIEW

**Comment by fire** [comment]
From Akien.

> So I think we need to keep them for compatibility. We can maybe keep them with #ifndef DISABLE_DEPRECATED, so we still have a way to remove them for updating the engine (and this should also be checked by CI, at least on the "Minimal template" build, so not covering the whole codebase).

This pull request was created because we want to keep backwards compatibility but we also need a way to check for `Ref/nullptr` comparisons in continuous integration and continuous delivery.

**Review by Repiteo — APPROVED** [review]
I think it's fine to also have `DISABLE_DEPRECATED` forbid the comparison, though I'm not sure if that's better handled in SCons or cpp. Either way, this implementation makes sense and the name is reasonable as-is.

File: .github/workflows/android_builds.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ on:
 env:
   # Used for the cache key. Add version suffix to force clean build.
   GODOT_BASE_BRANCH: master
-  SCONSFLAGS: verbose=yes warnings=extra werror=yes debug_symbols=no module_text_server_fb_enabled=yes
+  SCONSFLAGS: verbose=yes warnings=extra werror=yes debug_symbols=no module_text_server_fb_enabled=yes use_strict_ref_checks=yes
```
Comment (fire) [inline .github/workflows/android_builds.yml:9]: What do you think of using `check_reference_comparison=yes`?

The current naming is a bit cumbersome and ref is a shorthand.

Comment (fire) [reply .github/workflows/android_builds.yml:9]: Maybe `check_refcounted_comparison` or `check_ref_comparsion`.

Comment (AThousandShips) [reply .github/workflows/android_builds.yml:9]: Note that `ref` here refers to the `Ref` container, not `RefCounted` so `ref` is appropriate here IMO

What about `no_ref_null_compare`? Or as an inverted one defaulting to `true`, `ref_null_compare` or `ref_nullptr_compare`?

Comment (fire) [reply .github/workflows/android_builds.yml:9]: `ref_null_compare=yes` Sounds good to me.

**Comment by AThousandShips** [comment]
A `strict_dev_checks` would make sense, or `DEV_ENABLED`

**Comment by AThousandShips** [comment]
Pros and cons:
* `DISABLE_DEPRECATED`:
  - `+` Unobtrusive and no new options
  - `-` Unusable for CI unless we add at least one editor build to catch things outside the minimal build (which misses a large portion of the engine, beyond the editor code)
* `enum_inspector_improve`:
  - `+` Precise, gives you exactly what you want
  - `-` Clutters the options
* `strict_dev_checks`:
  - `+` Expandable
  - `-` Clutters the options (albeit less than above)
* `DEV_ENABLED`:
  - `+` Unobtrusive and no new options
  - `+` Covered by CI
  - `-` Less complete coverage (not all platforms are covered)

Will tentatively change to a more general option of `strict_dev_checks` as I think personally it's the cleanest option, and we can see what further discussion leads to

**Comment by Repiteo** [comment]
Thought about it more & realized there's another option: `DISABLE_DEPRECATED` OR `DEV_ENABLED`. That gives pretty significant overlap to most of our use-cases.
```diff
-#ifndef REF_NULL_COMPARE
+#if defined(DISABLE_DEPRECATED) || defined(DEV_ENABLED)
 	// Delete these to prevent raw comparisons with `nullptr`.
 	bool operator==(std::nullptr_t) const = delete;
 	bool operator!=(std::nullptr_t) const = delete;
-#endif // REF_NULL_COMPARE
+#endif // defined(DISABLE_DEPRECATED) || defined(DEV_ENABLED)
```
---
If we go with the generic option route, I'd much prefer something more generalized like `strict_checks` or even `strict`, as there's already the ambiguity of having both `dev_mode` and `dev_build`. It'd make more sense to have developer association be implicit by including it in `dev_mode`.

**Comment by Repiteo** [comment]
Similarly, we should probably add these deleted comparisons to `Variant` as well. While it's likely not nearly as abused, there's still that potential for not utilizing the `is_null()` function. Could be saved for a followup PR though.

**Comment by AThousandShips** [comment]
The `Variant` code is very different and the `is_null` method on `Variant` is not as reliable and safe unless you are looking for general purpose (for example it uses the generally unsafe `Callable::is_null`) so `is_null` for `Variant` is not equivalent to comparing with `nullptr`

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
> If we go with the generic option route, I'd much prefer something more generalized like `strict_checks` or even `strict`, as there's already the ambiguity of having both `dev_mode` and `dev_build`. It'd make more sense to have developer association be implicit by including it in `dev_mode`.

Yeah we can go with `strict_checks` and default it enabled for `dev_mode`.

**Comment by AThousandShips** [comment]
Will rename

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank *you*!

**Pull request merged**